modules/raft/src/main/java/org/apache/ignite/internal/raft/server/impl/JraftServerImpl.java

```diff
@@ -856,4 +860,54 @@ public void onLeaderStart(long term) {
             listener.onLeaderStart();
         }
     }
+
+    /** Returns raft group state accessor. */
+    public <G extends ReplicationGroupId, L> RaftGroupStateProvider<G, L> getGroupStateProvider(
+            Class<G> groupType,
+            Class<L> groupListenerType
+    ) {
+        return new StateProviderImpl<>(groupType, groupListenerType, nodes);
+    }
+
+    private static class StateProviderImpl<G extends ReplicationGroupId, L> implements RaftGroupStateProvider<G, L> {
```

Comment:
Still seems a bit too complicated to me.
Why not to introduce new service MinimumRequiredTimeCollectorService, instantiate it in IgniteImpl and propagate to TableManager -> PartitionListener, similar to what you've did with MinimumRequiredTimeProvider.
Internally MinimumRequiredTimeCollectorService will be a map partitionId -> minTxStartTime, where all partitionListeneners will put (localId, minTxStartTime) on handleUpdateMinimalActiveTxTimeCommand.

Seems simpler, though I can't say that it's beautiful.

Reply:
@sanpwc Sounds reasonable. But could you please clarify who is going to cleanup this map and when?

Reply:
I believe that a catalog compactor may clean the whole map on each successful compactisation.

Reply:
Removed modifications of raft-code in favour of  `MinimumRequiredTimeCollectorService `. that stores `TablePartitionId` -> `timestamp`
- When a partition is added it adds an entry to that map.
- its updates these entries when needed.
- When a partition is destroyed it removes its entry from the map.